reference-implementation/lib/transform-stream.js

```diff
@@ -229,25 +229,15 @@ class TransformStreamSink {
 
     assert(transformStream._transforming === false);
 
-    // No control over the promise returned by writableStreamWriter.close(). Need it?
-
     transformStream._writableDone = true;
 
-    if (transformStream._transformer.flush === undefined) {
-      TransformStreamCloseReadable(transformStream);
-    } else {
-      try {
-        transformStream._transformer.flush(
-            transformStream._enqueueFunction,
-            transformStream._closeFunction,
-            transformStream._errorFunction);
-      } catch (e) {
-        if (transformStream._errored === false) {
-          TransformStreamErrorInternal(transformStream, e);
-          throw e;
-        }
-      }
-    }
+    const flushPromise = PromiseInvokeOrNoop(transformStream._transformer, 'flush', [transformStream._controller]);
+    // Return a promise that is fulfilled with undefined on success.
+    return flushPromise.then(() => TransformStreamCloseReadableInternal(transformStream),
```

Comment:
You're right, that should be fixed, though I'm tempted to make `controller.error()` throw after `flush()` is called, like with `controller.close()`.

Otherwise, what should happen if controller.error() is called with one error, and flush throws or rejects the promise with a different error? What should happen if controller.close() is called, but then the flushPromise is rejected? It seemed simpler to give flush only a single way of signaling success or failure.